### PR TITLE
Add a new field to the Delivery Service to accommodate for "max_request_header_size" 

### PR DESCRIPTION
**Pull request by srijeet0406**
## What does this PR (Pull Request) do?

- [x] This PR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Control Client 
- Traffic Ops

## What is the best way to verify this PR?

Run the `goose` migrations for both dev and test
Make sure all the unit tests/ API tests pass
Try creating a new DS in TP, and make sure the `Max Request Header Size` field shows up.
Make sure you can also add and remove that field from the TP DS table view.
Try updating that field and make sure that the updates take effect.
## If this is a bug fix, what versions of Traffic Control are affected?

- master
## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR includes database migrations
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

## Additional Information

This PR is the prerequisite for ORT to complete https://github.com/apache/trafficcontrol/issues/5244
Once this is merged, ORT will be required to inject the header rewrite rules to enforce that the constraint is followed

### CODE REVIEW

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [inline lib/go-tc/deliveryservices.go:182]: Since 5.0 is leaving the station, we probably need a new `DeliveryServiceNullableV31` struct and all the other changes that come with that, since this will likely be 3.1. Right @ocket8888? This shouldn't be backported.

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:862]: The way the DS API works, we also need to query the existing value of `max_request_header_size` in `updateV15`, in order for 1.5 requests to be properly _upgraded_ to 3.0 requests. Otherwise, a pre-3.0 request would blow away the existing value.

The [README](https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/README.md#updating-a-minor-version) explains this a bit better, but might be out of date now that `DeliveryServiceNullable` is deprecated :(

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: lib/go-tc/deliveryservices.go
Comment (ocket8888) [reply lib/go-tc/deliveryservices.go:182]: Well hopefully the next release can include the DS `active` flag change, so we may actually want to go to 4.0. That'll be a breaking change that needs a new major version bump.

Comment (srijeet0406) [reply lib/go-tc/deliveryservices.go:182]: The DS `active` flag change(that I worked on) is not modifying the structure. It just lets us query by the `active` flag.

Comment (rawlinp) [reply lib/go-tc/deliveryservices.go:182]: He's referring to https://github.com/apache/trafficcontrol/blob/master/blueprints/ds-active.md which hasn't been started yet. That blueprint would call for a major version bump.

_This_ PR could add it as 3.1, then if we get to ☝️ before 5.1 is released, that API could still be added as 4.0. Then 3.1 == 4.0 like how 1.5 == 2.0 (if that makes sense). It _would_ be cleaner if we were able to have just one or the other, but I don't think it's a big deal if we end up having to add both.

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [inline lib/go-tc/deliveryservices.go:197]: So I had some thoughts @ocket8888 @rob05c. I think we may want to continue using type aliases like this, but only per major version.

Otherwise, let's say we end up with another 5 minor versions in 3.x. That means we will have 5 different `GetDeliveryServiceV*`, `UpdateDeliveryServiceV*`, etc, all in the same client package. I don't think that's what we want the client to look like, right? As a user, I wouldn't want to have to pick between all those minor version methods. I thought we wanted it to auto-downgrade minor versions, which means all the client functions should be using a type alias per major version, e.g. `type DeliveryServiceV3 DeliveryServiceV31`, which we continually update to alias to new minor versions, e.g. `type DeliveryServiceV3 DeliveryServiceV32`. Does that make sense?

Since we already have `DeliveryServiceNullableV30`, maybe that should be the current alias. Then we don't need to add new client methods for V31.

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [inline lib/go-tc/deliveryservices.go:352]: I think we might not need this if we alias `type DeliveryServiceV30 DeliveryServiceV31` like I mention in the previous comment. But if we don't want to do that for some reason, I think the matching `DeliveryServiceV30` methods would no longer be in use and could be removed. We should only be calling `Sanitize`, `Validate`, etc. in the _latest_ handlers.

File: traffic_ops/app/db/migrations/2020113000000000_add_max_request_header_size_delivery_service.sql
Comment (rawlinp) [inline traffic_ops/app/db/migrations/2020113000000000_add_max_request_header_size_delivery_service.sql:22]: missing newline at end of file

File: traffic_ops/testing/api/v3/deliveryservices_test.go
Comment (rawlinp) [inline traffic_ops/testing/api/v3/deliveryservices_test.go:74]: We shouldn't need new minor-version specific tests if we follow my first comment above. We could just check this new field in the existing `GetTestDeliveryServices` test (similar for update).

File: traffic_ops/testing/api/v3/traffic_control_test.go
Comment (rawlinp) [inline traffic_ops/testing/api/v3/traffic_control_test.go:34]: We probably wouldn't need this and the new additions to tc-fixtures.json either (assuming we alias like I mention in the first comment).

File: traffic_ops/traffic_ops_golang/routing/routes.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/routing/routes.go:148]: We should only need 3.1 routes for POST and PUT. The GET handler should have the minor-version specific handling, and DELETE is version-agnostic.

File: traffic_ops/v3-client/deliveryservice.go
Comment (rawlinp) [inline traffic_ops/v3-client/deliveryservice.go:40]: If we follow my aliasing comment above, I think we would just update `apiBase` to 3.1, and we wouldn't need the new 3.1-specific client methods below.

File: traffic_portal/app/src/common/api/DeliveryServiceService.js
Comment (rawlinp) [inline traffic_portal/app/src/common/api/DeliveryServiceService.js:23]: I think we would just update `ENV.api['root']` to 3.1, right @mitchell852? I think we would prefer to have TP using the same version for all requests if possible.

**Review by rawlinp — CHANGES_REQUESTED** [review]
I'd like to get some feedback from @ocket8888 and @rob05c on this as well. We may need to rethink the whole "deprecate the `DeliveryServiceNullable` alias" thing.

File: lib/go-tc/deliveryservices.go
Comment (TaylorCFrey) [inline lib/go-tc/deliveryservices.go:388]: Is there something special about this value (131072)? Should we asign it as a variable to avoid magic number confusion in the future?

File: lib/go-tc/deliveryservices.go
Comment (TaylorCFrey) [inline lib/go-tc/deliveryservices.go:35]: Are these max values used?

File: lib/go-tc/deliveryservices.go
Comment (TaylorCFrey) [inline lib/go-tc/deliveryservices.go:191]: Do we need to signal if this, too, is `nullable` or not, like the nested struct?

File: lib/go-tc/deliveryservices.go
Comment (TaylorCFrey) [inline lib/go-tc/deliveryservices.go:638]: Would these also be considered _"magic numbers"_ even though they are explained in comments?

File: lib/go-tc/deliveryservices.go
Comment (ocket8888) [reply lib/go-tc/deliveryservices.go:197]: That should work out fine, since the only kinds of changes that should break alias property access are the same kinds that would necessitate a major version bump.

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [reply lib/go-tc/deliveryservices.go:191]: I believe the general rule is that new structs moving forward should be nullable, so it's not really necessary to call that out for new structs anymore.

File: traffic_portal/app/src/common/api/DeliveryServiceService.js
Comment (ocket8888) [reply traffic_portal/app/src/common/api/DeliveryServiceService.js:23]: That's a possibility, but it will mean needing to add routes to `routes.go` for 3.1 for all API routes.

Comment (rawlinp) [reply traffic_portal/app/src/common/api/DeliveryServiceService.js:23]: That's not the way routes work unless it was changed. If there is a single 3.1 route in existence, any handlers for 3.0 will implicitly handle 3.1 requests.

File: lib/go-tc/deliveryservices.go
Comment (srijeet0406) [reply lib/go-tc/deliveryservices.go:388]: This is the default value used for our header sizes, I have made it into a constant.

File: lib/go-tc/deliveryservices.go
Comment (srijeet0406) [reply lib/go-tc/deliveryservices.go:638]: Made constants for these values as well.

File: lib/go-tc/deliveryservices.go
Comment (TaylorCFrey) [reply lib/go-tc/deliveryservices.go]: Cool. I meant is there anything unique about 131072 vs, say, 32768? Is that bits or bytes? According to https://stackoverflow.com/questions/686217/maximum-on-http-header-values it looks like the max header size is 4K to 8K, but it depends on the underlying server or page files, etc. Not sure if this that is even relevant.

File: lib/go-tc/deliveryservices.go
Comment (rob05c) [inline lib/go-tc/deliveryservices.go:194]: Would you object to changing this to `MaxRequestHeaderBytes`?
I think it makes it easier, so people don't have to look up the documentation to know the units, and don't assume the wrong unit and make a mistake.

File: lib/go-tc/deliveryservices.go
Comment (rob05c) [reply lib/go-tc/deliveryservices.go:200]: FWIW, I ended up making an alias in `lib/go-atscfg` for both the latest DS and Server. Just for stability, because it kept changing, and requiring updating hundreds of lines of code. Where using the alias lets me just update that single location, and the few places that broke.

I don't currently have a strong opinion on this. But I think we should do whatever is easy for users to work with. From my recent experience, that seems to be an alias.

> That should work out fine, since the only kinds of changes that should break alias property access are the same kinds that would necessitate a major version bump.

Right, that's the big issue: how does an updating alias work with Go Modules and the library Version Promise?
It's a little bit nebulous what constitutes "breaking." The updating minor alias will only break if people use the nested initialization, e.g. `sv := tc.Server{tc.ServerV3{tc.ServerV2{Foo: foo}, Bar: bar}}` vs `sv := tc.Server{}; sv.Foo = foo; sv.Bar = bar`.

I think I'd be ok with declaring that unsupported in terms of the Go Modules and the library compatibility promise, and documenting so a symbol comment. "Struct alias initialization is only guaranteed forward-compatible for default-initialization."

Comment (rawlinp) [reply lib/go-tc/deliveryservices.go:200]: That sounds good to me.

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:874]: This is kind of confusing to review since it seems like this was all added as new code, but I think things were just moved around.

For the update* functions, you should just rename `updateV30` to `updateV31`, and change it to take in `DeliveryServiceV31` as input and output. Then _within_ the function you should immediately cast it to `DeliveryServiceNullableV30`. At the end of the function, you cast it back to `DeliveryServiceV31` before returning it.

Then you can basically copy `updateV15` to `updateV30`, change it to take `DeliveryServiceV30` as input and output, then change it to read the new `max_request_header_bytes` from the DB before passing it to `updateV31`, where the real handling is done.

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (srijeet0406) [reply traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:874]: Yeah this is not a new change, it was just moved around.

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [inline lib/go-tc/deliveryservices.go:448]: We should do some basic validation for `MaxRequestHeaderBytes` as well -- like it should be greater than zero.

File: lib/go-tc/deliveryservices.go
Comment (rawlinp) [inline lib/go-tc/deliveryservices.go:35]: we should rename this `DefaultMaxRequestHeaderBytes` to match the new name

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:274]: We should rename `ds` to `dsV13`, then we can use `ds` instead of `dsNullableV30` on L278. Essentially, we don't want to be distracted by the versioning stuff throughout the method. Also it cuts down on the number of lines we have to change when adding new versions.

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:761]: So this function should actually stay exactly the same as it was before. Basically, we want to rename `updateV30` to `updateV31`, change it to accept and return V31 structs, then copy `updateV15` to `updateV30`, change it to accept V30 structs, upgrade them to V31 by changing the query to only select the new V31 field from the DB, then pass it to `updateV31` for the actual handling.

It's kind of hard to draw in a review, but the functions are designed to chain together like so:
```
UpdateV12 -> updateV12
              |     ^
              V     |
UpdateV13 -> updateV13
              |     ^
              V     |
   ...          ...
UpdateV31 -> updateV31  <------(this latest handler does all the actual handling)
```
So each API version is like a link in a chain. For adding a new version, you just add a new link to the end of the chain, and you don't have to worry about earlier links in the chain (other than the one you're linking to). Does that make sense?

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:800]: similar to my comment for `createV31`, we should rename this parameter to `dsV31`, then use `ds` instead of `dsNullableV30` on L803

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:1030]: Per my earlier comment above, this function should look just like `createV30`, except with the DB query that reads the existing value of the new field.

**Review by rawlinp — CHANGES_REQUESTED** [review]

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:800]: So I figured out why the test for this wasn't failing and am about to open a PR. This should be added back.

Comment (srijeet0406) [reply traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go:800]: Whoops sorry, I must have fat fingered a copy paste. fixed now.

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.DNS.tpl.html
Comment (rawlinp) [inline traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.DNS.tpl.html:406]: this should probably say "The maximum size (in bytes) of the ..." (as well as the tooltip on the other form)

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.HTTP.tpl.html
Comment (mitchell852) [inline traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.HTTP.tpl.html:411]: the value attribute is not needed here. if this is a new ds, the value of this field will be derived from deliveryService.maxRequestHeaderBytes that was set in traffic_portal_properties.json. if this is an existing ds, it will show the ds's value.

**Review by mitchell852 — CHANGES_REQUESTED** [review]
looks good. one thing however.

File: traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.HTTP.tpl.html
Comment (srijeet0406) [reply traffic_portal/app/src/common/modules/form/deliveryService/form.deliveryService.HTTP.tpl.html]: Made the change

**Comment by mitchell852** [comment]
sorry if this was already discussed but was it considered using a value of 0 to mean "no max"? We have a precedent with that. See max dns answers and max origin connections.

**Comment by srijeet0406** [comment]
@mitchell852 No, 0 is considered an invalid value. It wont let you create a DS with the max header bytes set to 0 in TP or in the backend TO.

**Comment by mitchell852** [comment]
> @mitchell852 No, 0 is considered an invalid value. It wont let you create a DS with the max header bytes set to 0 in TP or in the backend TO.

honestly, i don't like 0 to mean "no max" but it's kinda weird that some properties on a ds, you can use 0 and others you cannot but i'll defer to you guys.

**Comment by rawlinp** [comment]
Hmm... I suppose if 0 meant "no max", that would most likely be a lie, because by default ATS sets the max to `131072` (which is why we're also setting that as the default here).

**Comment by srijeet0406** [comment]
> > @mitchell852 No, 0 is considered an invalid value. It wont let you create a DS with the max header bytes set to 0 in TP or in the backend TO.
> 
> honestly, i don't like 0 to mean "no max" but it's kinda weird that some properties on a ds, you can use 0 and others you cannot but i'll defer to you guys.

Yeah, 0 header size would just mean that you cannot add a header rewrite rule in there, which is not what we want. You will always have to specify a value > 0.

**Review by rawlinp — APPROVED** [review]
Code looks good. Automated tests passed. Manual testing worked. 👍

**Comment by mitchell852** [comment]
@srijeet0406 - since you changed traffic_portal_properties.json, you'll need to make a change like this to account for the new field: https://github.com/apache/trafficcontrol/pull/5379/commits/a1e504c6468f679d077c1192821ab498b0899be5

**Comment by mitchell852** [comment]
image

Think you could expand this to explain what leaving it blank/null will do?

**Comment by srijeet0406** [comment]
> image
> 
> Think you could expand this to explain what leaving it blank/null will do? For example, if empty is there a) no max or b) the global max (if defined) kicks in?

It actually cannot be empty, the min value of 1 will always be needed.

**Comment by rawlinp** [comment]
If you mean from the API perspective, it will sanitize a null value to the default.

**Comment by mitchell852** [comment]
> If you mean from the API perspective, it will sanitize a null value to the default.

Oh i see. so maybe the tooltip should say something like "leaving this value empty will result in the default: 131072". By the way what's the reasoning behind the 131072?

**Comment by mitchell852** [comment]
oh geez. i see...it can't be empty. i should actually read.

**Comment by mitchell852** [comment]
so i can clear the value

image

but then the default comes back so the tooltip should probably say that.

**Comment by mitchell852** [comment]
bug: I entered 2147483648 (1 higher than int range). i guess 2147483647 should be enforced as the max in UI/API

**Comment by srijeet0406** [comment]
> bug: I entered 2147483648 (1 higher than int range). i guess 2147483647 should be enforced as the max in UI/API

Fixed in the latest commit

**Comment by rawlinp** [comment]
> By the way what's the reasoning behind the 131072?
That's the default that ATS uses for this.

**Comment by mitchell852** [comment]
@srijeet0406 - isn't this related to #5244? want to add that to the description? also, want to share this with/demo this to @dsouza93 before merging to see if he has any feedback?

**Comment by mitchell852** [comment]
@srijeet0406 - sorry, this is not really related to your PR but can you make a tiny change for me? I found a small bug in table.serverCapabilityDeliveryServices.tpl.html. line 37 should be `<th>XML ID (Key)</th>` instead of `<th>Host</th>`. Think you can make that super tiny change?

**Comment by srijeet0406** [comment]
@mitchell852 No problem at all, I've fixed it in the latest commit. Also, I did have a talk with Dylan and showed him a quick demo earlier. He seemed happy with the changes. So if you feel it looks good, I'd say we are good here :)

**Comment by srijeet0406** [comment]
> @srijeet0406 - isn't this related to #5244? want to add that to the description? also, want to share this with/demo this to @dsouza93 before merging to see if he has any feedback? or I suppose it could be demo'd after merging but any changes would require a new PR. just let me know. this all looks good. i'm ready to merge on your request.

I've mentioned it in the "additional info" section of the PR

**Comment by TaylorCFrey** [comment]
> @srijeet0406 - sorry, this is not really related to your PR but can you make a tiny change for me? I found a small bug in table.serverCapabilityDeliveryServices.tpl.html. line 37 should be `<th>XML ID (Key)</th>` instead of `<th>Host</th>`. Think you can make that super tiny change?

FWIW I think this particular sized bug is fine to squeeze into a PR like this, but I do think a quick GH issue would be nice to ensure there was some level of tracking. Not necessarily a separate PR, but if the issue was tracked and included via reference in this PR then we should be good to go. What do you think @mitchell852?

**Comment by mitchell852** [comment]
> > @srijeet0406 - sorry, this is not really related to your PR but can you make a tiny change for me? I found a small bug in table.serverCapabilityDeliveryServices.tpl.html. line 37 should be `<th>XML ID (Key)</th>` instead of `<th>Host</th>`. Think you can make that super tiny change?
> 
> FWIW I think this particular sized bug is fine to squeeze into a PR like this, but I do think a quick GH issue would be nice to ensure there was some level of tracking. Not necessarily a separate PR, but if the issue was tracked and included via reference in this PR then we should be good to go. What do you think @mitchell852?

normally i'd agree but this is so tiny/cosmetic seems like a waste of time to create a new issue/PR and really tracking this level of change is probably unnecessary and provides almost no value.

File: traffic_ops/app/db/migrations/2020113000000000_add_max_request_header_size_delivery_service.sql
Comment (rawlinp) [inline traffic_ops/app/db/migrations/2020113000000000_add_max_request_header_size_delivery_service.sql:1]: Another migration has been merged, so this date will need to be updated to be later than the recently merged one.

**Review by mitchell852 — APPROVED** [review]
all looks good to me!

**Pull request merged**
